### ISSUE TITLE
Make the MultiBlock API saner

### ISSUE THREAD

**Issue by adeak** [pyvista/pyvista#1507]
This is just to track our discussion in https://github.com/pyvista/pyvista/pull/1482#discussion_r664009578:

> This is a pretty confused API :(
> 
>   1.  `blocks.keys()` is a thing, and instantiating from a list comes up with string keys, so it looks like a dict.
>   2.  Iterating over the container, however, yields values, which is not like a dict.
>   3.  There's no `blocks.values()` nor `blocks.items()`...
>   4.  `blocks[i]` supports both keys and integer indices?!

Additionally, I've also noticed that `MultiBlock__setitem__` does this:
```py
import pyvista as pv

blocks = pv.MultiBlock()
blocks[3, 'name'] = pv.Sphere()

blocks.keys()
# [None, None, None, 'name']
```

Making a 2-tuple passed to `__setitem__` act neither like a list nor like a dict is especially problematic from an API design standpoint.

I'm still not sure if we can change any of this without breaking too much downstream code, but at least we can track this future work on this issue.

**Comment by adeak** [comment]
Perhaps deprecate this too if we touch this: https://github.com/pyvista/pyvista/blob/e79c03369dcaf57292ae6b7b366410d82debcee7/pyvista/core/composite.py#L434-L443

If we handle the MutiBlock as an iterator then it should just rely on `next(multi)` rather than exposing a `multi.next()` that's synonymous. Having a `.next()` method on an iterator is a python2ism.

**Comment by akaszynski** [comment]
Agreed.  Deprecate `next`.

Would it make sense to have `MultiBlock` look like an OrderedDict with bonus pyvista features?

**Comment by adeak** [comment]
Hmm. It should probably be more like a SortedDict (which is not to say that I know of an implementation of such). We want users to be able to insert in the middle of defined keys, which is different from the behaviour of an OrderedDict (or python 3.7+ dict). Perhaps an OrderedDict plus a private list of names can work. That would have the added benefit that `multi.clean` would just change that one list without touching any of the data.

But I strongly feel that it should be some kind of dict, with list-like bells and whistles added. We'd have to think hard about making the above insertion sane before committing to a new, similarly weird API :) We should start by agreeing on a set of behaviours that we think is consistent (and pythonic). I'll think about this for a bit.

**Comment by MatthewFlamm** [comment]
MultiBlock meshes in vtk have direct support for indices.  The string identifiers are in the metadata.  I would suggest that pyvista adopt this as well.  That is MultiBlock meshes are list like.  There would be a separate API to interact via string keys.  Something like:

```
block = pyvista.MultiBlock([mesh1, mesh2, ...])
block.names = ["mesh_name1", "mesh_name2", ...]
block.get("mesh_name2")
block[2]
```

**Comment by adeak** [comment]
> That is MultiBlock meshes are list like.

That would also solve some vague qualms I've been forming with iteration behaviour.

**Comment by MatthewFlamm** [comment]
Looking more into the vtk implemention https://vtk.org/doc/nightly/html/classvtkMultiBlockDataSet.html#aa628665852cd6d5a5f99afb6fc16cfe1

> It is recommended that one uses the iterators to iterate over composite datasets rather than using this API.

For MultiBlock meshes, this seems like a suggestion rather than a requirement.

Other composite datasets require using the vtk iterators to access and set data.  For example: https://vtk.org/doc/nightly/html/classvtkUniformGridAMR.html

I'm pointing this out in case it makes sense to have a higher level view for this.  I'm not sure how to handle it myself.

**Comment by adeak** [comment]
> > It is recommended that one uses the iterators to iterate over composite datasets rather than using this API.
>
> For MultiBlock meshes, this seems like a suggestion rather than a requirement.

Without having checked the implementation, that sounds a lot like "list-like". Iteration for lookup means O(N), so this must mean that `GetBlock()` would do the same thing anyway under the hood. So based on that note I would expect a single `GetBlock()` call to be the same as iterating, and repeated sequential `GetBlock()` calls to be much worse (O(N^2) overall).

**Comment by MatthewFlamm** [comment]
> Without having checked the implementation, that sounds a lot like "list-like".

I agree.  If we implement the list-like suggestion as the public api, then we should be able to implement either method inside.

**Comment by MatthewFlamm** [comment]
Coming back to this as I got bitten again recently by this unintuitive API.  I can think of two separate use cases: 

One is when reading data from a file format like OpenFOAM where there is a nested multiblock structure.  Here there user doesn't usually know which order the mesh objects are put in, and instead rely on the names.  This would suggest a dict like implementation.

The second is when algorithmically creating many objects, possibly identical ones.  Like in #2941.  This would suggest a list implementation.

I propose we should choose one primary implementation and provide useful methods to allow some functionality for the secondary. I propose a list-like implementation, since vtk allows duplicate names.  More specific proposal:

1. A list-like primary implementation, e.g. `multi[1] = mesh`
2. When possible, add optional args for setting key, e.g. `def append(mesh, name=None)`
3. dict-like methods: `get` and `set` that take the place if the `__setitem__` and `__getitem__`.
  * It could be confusing since `get` is different from `__getitem__` in a dict.  Do we raise if `key` does not exist?
4. We still preserve the dict structure when doing operations with the list API, i.e. preserving the key, value pairs when reversing the list.

Currently we set default key names if none is provided.  One downside of this approach is that this won't be equal if we set default names/keys, since we check for the same keys in the equality. This might suggest removing default key names.

```py
a = pv.MultiBlock([pv.Sphere(), pv.Cube()])
b = pv.MultiBlock([pv.Cube(), pv.Sphere()])
assert a == b.reversed()
```

Another downside to a strict list-like implementation is that it could be hard to set the value and key at the same time.  Currently we can do `multi[5, "new_name"] = mesh`.  Point 2 above tries to minimize this problem, but doesn't eliminate it.  I suppose we could continue to allow this usage as it is opt-in.

**Comment by adeak** [comment]
> 1. A list-like primary implementation, e.g. multi[1] = mesh

Works for me.

> 2. When possible, add optional args for setting key, e.g. `def append(mesh, name=None)`

Sounds reasonable.

> 3. dict-like methods: get and set that take the place if the `__setitem__` and `__getitem__`.
> * It could be confusing since get is different from `__getitem__` in a dict. Do we raise if key does not exist?

How about implementing a dict-like `get`? `multi.get(key)` raises if missing, `multi.get(key, default)` returns `default` if missing. Allows to write pythonic code, looking like a dict.

I wonder, should we allow `__getitem__`/`__setitem__` to choose based on type? If we have a strict list-like structure (int indices) and optional names (string names), we could trigger dict-like lookup based on key type.

> Currently we set default key names if none is provided. One downside of this approach is that this won't be equal if we set default names/keys, since we check for the same keys in the equality. This might suggest removing default key names.

I don't have an opinion about this either way. If we go with "list with dict bells" rather than "dict with list bells" then it's not an issue to have `None` default keys.

> Another downside to a strict list-like implementation is that it could be hard to set the value and key at the same time. Currently we can do `multi[5, "new_name"] = mesh`. Point 2 above tries to minimize this problem, but doesn't eliminate it. I suppose we could continue to allow this usage as it is opt-in.

Honestly I don't like this syntax to begin with (but I don't use `MultiBlock`s so this is not a strong signal). I'd be more tempted to allow 2-tuple assignments: `multi[5] = "new_name", mesh`. Still opt-in, and this reflects the "list containing dict items" kind of logic. Not prettier, I admit, but looks more consistent to me. That being said, I'm open to being pragmatic here, because as you said this is opt-in.

**Comment by MatthewFlamm** [comment]
> Works for me. What happens if someone assigns a missing index?

My proposal is that we stick to list like implementation whenever possible, so raise `IndexError`.

> How about implementing a dict-like `get`? `multi.get(key)` raises if missing, `multi.get(key, default)` returns `default` if missing. Allows to write pythonic code, looking like a dict.

Makes sense to me.

> I wonder, should we allow __getitem__/__setitem__ to choose based on type? If we have a strict list-like structure (int indices) and optional names (string names), we could trigger dict-like lookup based on key type.

I like this too.  It is a nice sugar on top.

> I'd be more tempted to allow 2-tuple assignments: multi[5] = "new_name", mesh

I like this better than the current API too. We could also consider allowing a dict to be passed in, since it would align with the secondary API usage.

**Comment by MatthewFlamm** [comment]
A list of 'features' of the current API that are surprising if the expected behavior is either list-like or dict-like, without trying to provide a value judgement on them.  I will propose a clean up and removal of a lot of these features in a PR, and I want to track them for discussion purposes.

- `multi[-1] = data`  # appends data
- `multi = MultiBlock(); multi[5] = data`  # sets length with an out of bound index
- `multi[1, 'name'] = data`  # sets key of data using second item of tuple `index` in `__setitem__`
- `multi[['name1', 'name2']]`  # use lists of str instead of slices
- `multi[[0, 1]]`  # use lists of int instead of slices

The last two I don't believe are currently supported for `__setitem__` only `__getitem__`

**Comment by MatthewFlamm** [comment]
You are right, I'm focusing first on `__setitem__`/`__getitem__` in this list.  I wanted to start writing down the current behavior in one place.  Feel free to add directly to my list or start a new list for other behaviors.
